### ISSUE TITLE
`playback` function

### ISSUE THREAD

**Issue by tim-salabim** [r-spatial/mapedit#37]
@timelyportfolio I really like the idea of the flubberized playback. In a scenario where person A does:
```r
library(mapedit)
library(mapview)
edited = editFeatures(franconia[1, ]) # do some editing
save(edited, file = "edited.rda")
```
and then sends `edited.rda` to person B, person B could then ideally simply use:
```r
load("edited.rda")
playback(edited)
```
to get a quick impression of what has been done by person A. 

Also, I think this would be a great show-off for the conference talk :-)

**Comment by timelyportfolio** [comment]
I agree this is nice functionality and makes for a good demo.  Beyond flubberizing, I could enivison some sort of `diff` map, using colors, alpha, and other styling to demonstrate changes.  Not sure we will have time for a full implementation by useR but something we could explore as a prototype.

**Comment by tim-salabim** [comment]
Yip, prototype sounds like a great idea! Even one that only does one thing would be enough. It may just spark enough feedback to collect for further implementation.

**Comment by timelyportfolio** [comment]
I will need to think through `add` and `delete` in a flubber context.  Also will need to make sure I join properly.

**Comment by tim-salabim** [comment]
There's no rush. This is surely not a must-have but a nice-to-have feature

**Comment by timelyportfolio** [comment]
We could easily code up a simple multi-user edit session also with Shiny.

**Comment by tim-salabim** [comment]
I am not exactly sure what you mean?

**Comment by timelyportfolio** [comment]
Sorry if I wasn't clear.  I meant we could we make a shared editing session in which we collaboratively edit a map.

Try this https://bl.ocks.org/timelyportfolio/cd81d81edf917dc711cfc4b2b4aa8bbf out as a little demo.  Will need to work out base map and a lot more, but hopefully gets point across of playing back an `editMap()`.

**Comment by tim-salabim** [comment]
In case we could wrap that in a function, say `playback`, that takes a `sf` object, checks for `recorder` attribute and if present returns such an animation, I think that would be great as a prototype even without any map background for starters.
I am still not sure what you mean by "a shared editing session in which we collaboratively edit a map"? I am assuming you mean recording of applied changes across multiple editing sessions by potentially multiple people in sequence? Or do you mean a session where multiple people work simultaneously? I wouldn't see a good use-case for the latter, but the former is definitely something that has value.

**Comment by timelyportfolio** [comment]
I meant the latter, and even if there was a lot of value, the infrastructure involved would make it out of scope for `mapedit`.  I just thought it would make for a nice demo, but given the usual wi-fi constraints at a conference, I agree it is not worth it.

In terms of multi-session sequential history, I wonder if we should wrap the return list with `recorder` attribute as a `s3`.  However, the `leaflet` ids would not sync, so might get tricky.

**Comment by tim-salabim** [comment]
For now I guess the attribute approach is good, as it still preserves the `sf` class. If we later on find that this is something that people really use very often then it will make sense to re-think the whole approach but for now I think the `recorder` implementation is fine as it is. If a prototype playback function is not too much of a hassle, then I think it is well worth providing it (maybe unexported), if it is a lot of work then I wouldn't put too much focus on.

**Comment by timelyportfolio** [comment]
I would prefer not to add a `d3r` dependency or add local d3 JavaScript in the package.  I could write unexported and then just insert remote JavaScript in `tags$head(tags$script())`.  That should suffice.  Let me know if you see problems with this.  I will add momentarily.

**Comment by timelyportfolio** [comment]
I just realized we also need `geojsonio`.  Instead of adding as a dependency, I will check with `requireNamespace` and bail with warning if `geojsonio` not available.

**Comment by tim-salabim** [comment]
Is this approach CRAN conform?

**Comment by timelyportfolio** [comment]
Yes, it has passed CRAN in other packages I have made.  Adding `geojsonio` to enhances should do the trick and not require a `mapedit` user to install `geojsonio`.

**Comment by tim-salabim** [comment]
Sweet, then I think this is a good approach. A suitable warning that for this particular piece `geojsonio` is needed should suffice.

**Comment by timelyportfolio** [comment]
What `projection` should we assume?  mercator?

**Comment by tim-salabim** [comment]
Yes, I guess web-mercator makes sense as things will look similar to what's been edited on the map.

**Comment by timelyportfolio** [comment]
Also, if we wanted to get real fancy, we could take this in many directions, but of course will require time, and I am guessing this is lower priority.  

* We could add to a leaflet map.
* We could provide something like a marble diagram ( see [example](http://rxmarbles.com/) ) for more granular playback.

**Comment by tim-salabim** [comment]
I think for now (i.e. useR!) it would be enough to highlight that we have something in the pipeline in this direction and show a first (even alpha-state) implementation. Hopefully we will this way gather some feedback and then take things from there. So, a rough function that animates changes on a white background is enough for now in my opinion.

**Comment by timelyportfolio** [comment]
Ok, I think I can get on leaflet in < 1 hour, so going to try to do really quick poc.

**Comment by timelyportfolio** [comment]
added both playback to pull #38

**Comment by timelyportfolio** [comment]
just made bug fixes and submitted + merged pull.  Let me know how it works.  Thanks!

**Comment by tim-salabim** [comment]
For both playback functions I only see the stuff draw in the first instance but there is no animation of any subsequent edits (i.e. moving vertices around)

**Comment by timelyportfolio** [comment]
I should have warned that for some reason `d3v4` does not work well with RStudio Viewer.  I will change to open in browser window.

**Comment by tim-salabim** [comment]
I see! Confirmed, it works in the breowser. Thanks, this is awsome!

**Comment by tim-salabim** [comment]
could this also work for `editFeatures`?

**Comment by timelyportfolio** [comment]
If I understand correctly, I think so, but we would have to know the original `sf` maybe as argument to the playback function.  I don't think it would be good to store the whole original `sf` in our `recorder` attribute.  I will try to think through the join in the JavaScript function also.

**Comment by timelyportfolio** [comment]
@tim-salabim going to remove `playback` and just pursue `playback_lf` as `playback` unless you have objections.  Also, getting close to a version sort-of-working with `editFeatures()` recorded session.

**Comment by tim-salabim** [comment]
No objections! It is definitely desirable to have the map background. Ideally, as in `editMap`, a user-defined map.

**Comment by tim-salabim** [comment]
Re playback with `editFeatures` could we utilise the `@object` slot in `mapview`?

**Comment by timelyportfolio** [comment]
Sorry, I am not sure I understand.  Would you mind clarifying?

**Comment by tim-salabim** [comment]
If we use `m = mapview(x)` internally to set up the map in `editFeatures(x)` then `m@object` will be a list with all sf objects displayed on the map (already projected to `crs = 4326`. Just a thought as you mentioned earlier that "we would have to know the original sf maybe as argument to the playback function".

**Comment by tim-salabim** [comment]
Ok, one last thing that would be really nice in playback (especially for `editFeatures`) is a grey static version of the original feature, so that comparisons to the original are made easy. Sorry for coming up with these things one after the other, but this is something that I am only exploring as we go along...

**Comment by timelyportfolio** [comment]
Great idea, I will try this out.  Here is where mapview slot might be helpful I think.
